Allow plugins to bust the cache.

rwjblue opened the pull request:
This is a small step towards fixing https://github.com/babel/broccoli-babel-transpiler/issues/67.

- [x] Needs tests
- [x] Verify with @stefanpenner on the naming of the cache key method for plugins
- [x] Basic Babel 6 support

#### Future:

- [x] Backport to 5.x branch (#94)
- [ ] Verify full babel 6 support with @Turbo87

rwjblue commented:
Updated to add some fairly simple caching tests.

stefanpenner commented:
It would also be useful to provide an array of paths to hash-for-dep, or maybe its sufficient for our plugins to hash-for-dep themselves?

rwjblue commented:
Yeah, I was thinking this is a good hook, and then we can provide a base plugin for our ecosystem that would have a default implementation that uses `hash-for-dep`.

stefanpenner commented:
@rwjblue the important bit is, plugins `MUST` not forget, so having such a API make encourage the baseline requirement.

rwjblue commented:
Updated to add more tests (identified 2 more scenarios that were previously uncovered).

@stefanpenner - Can you list what needs to be done to land this?  I believe it is somewhat urgently needed (I personally am forced to screw with caching a few times a week).

stefanpenner commented on index.js:
```diff
@@ -111,6 +111,28 @@ Babel.prototype.optionsHash = function() {
+        } else if (pluginAugmentsCacheKey) {
+          hash.plugins.push(plugin._augmentCacheKey());
+        } else {
+          hash.plugins.push(pluginType === 'function' ? plugin + '' : plugin);
```
I don't believe this case is valid, as the toString of a function is not sufficient to build a cache key, unless the function is closed and does not delegate to any external ode.

stefanpenner replied:
I also believe if a plugin is encountered, where we can't get a cacheKey we should throw, or warn or something.

rwjblue replied:
Ya, but we have to add *something*. I was simply emulating what was being done above for the main options keys.

stefanpenner replied:
> Ya, but we have to add something.

I don't believe so, whats the point if the cache key is not accurate? We should warn/throw/disable persistent cache if we cannot get a reasonable key.

rwjblue replied:
Because we need to introduce this and allow folks creating plugins to have a chance to update. 

I think a warning would be fine.

stefanpenner replied:
> We should warn/throw/disable persistent cache if we cannot get a reasonable key.

rwjblue replied:
Falling back to non-persistent mode (with a warning) seems great. But I'm unsure if you can downgrade after the constructor. 

I can play with that to see. If we can't change strategy then a warning + a completely generated key (to ensure it is unique and prevent long term persistence) would work.

stefanpenner replied:
persistent output is not configurable, or shouldn't be. But the cache strategy is easy to swap.

stefanpenner commented on index.js:
```diff
@@ -111,6 +111,28 @@ Babel.prototype.optionsHash = function() {
     }
 
+    if (options.plugins) {
+      hash.plugins = [];
```
can we have plugins provide a path to resolve from to find them and their dependents. This would allow babel-transpiler to hash-for-dep them.

rwjblue replied:
Sure that seems like a good addition to what is here, but I don't think it replaces the mechanim/concept in this PR already (since the plugins config needs to be taken into account).

stefanpenner replied:
In every case, the hash-for-dep of a plugin's npm module is what we need. What this PR adds, would be a complement to the hash-fpr-dep approach, not the other way around.

stefanpenner replied:
Most likely, we want to tell babel-tranpsiler to:

1. load a plugin name X relative to Y
2. provide it a mechanism to instantiate and configure that plugin, where the outputs of the config are automatically serialized and used to derive a cache key. If the serialization is is not possible, we should warn/throw/disable the cache.

rwjblue replied:
> What this PR adds, would be a complement to the hash-fpr-dep approach, not the other way around.

How is this a useful point? Both things are required, I have already agreed with that.

stefanpenner replied:
We discussed this out of band, hash-for-deps implementation was not clear (likely because it is not well documented). After hash-for-deps value was cleared up, its importance made sense. Yes both techniques should exist, and will work well together.

rwjblue commented:
Had a great chat with @stefanpenner about the path forward here.  Will try to get this updated today, and see if we can land something soon.

rwjblue commented:
OK, just updated.  At this point a plugin may provide a `baseDir` and `cacheKey` methods to be used.  If `baseDir` is not included a console warning is issued, and we fall back to non-caching mode.

@stefanpenner - I think this addresses the various things that we addressed, please let me know what you think...

stefanpenner commented on index.js:
```diff
@@ -111,6 +116,44 @@ Babel.prototype.optionsHash = function() {
+          // prevent caching completely if the plugin doesn't provide baseDir
+          // we cannot ensure that we aren't causing invalid caching pain...
+          this.console.log('broccoli-babel-transpiler is opting out of caching due to a plugin that does not provide a caching strategy: `' + plugin + '`.');
+          hash.plugins.push((new Date).getTime() + Math.random());
```
TODO: https://github.com/stefanpenner/broccoli-persistent-filter/issues/74

stefanpenner replied:
I think we can also early exit from the loop here.

stefanpenner replied:
lets be sure the two numbers don't merge, either use only `Math.random()` or ->  `(new Date).getTime()+ '|' + Math.random();`

stefanpenner replied:
What happens in the `else` case here. It is a-bit hard to reason about.

rwjblue replied:
There is no else, so either it's a string plugin (and takes no options) or it has a baseDir. Anything else gets the non-persistent treatment.  

Also, RE: the issue in b-p-f I'm happy to work on that next. I agree switching back out of persistent strat is better than what I'm doing here.

rwjblue replied:
I will switch to make sure that the numbers done merge (by coercing to strings), but when I wrote it I thought it might be good actually. Because `getTime` always returns an integer, and `Math.random()` always returns a float from 0 to 1 we would just end up with a value like `1444566677.129387273`.

stefanpenner commented on index.js:
```diff
@@ -111,6 +116,44 @@ Babel.prototype.optionsHash = function() {
+        if (pluginType === 'string') {
+          hash.plugins.push(plugin);
+        } else if (!pluginProvidesBaseDir){
+          // prevent caching completely if the plugin doesn't provide baseDir
```
it seems like if `plugin` is an object, but all its members are primitive we can just serialize it as well. (i think)

rwjblue replied:
I don't think that is legal (though I will have to check), I think (in 5.x) they need to be functions that are invoked?

rwjblue commented:
Updated based on last round of feedback.  Test suite now passing (had a bizarre quirk with the fixture files and `hash-for-dep`).  README has been updated.

I'd like to get final review on this so that I can land and release as soon as I have confirmed that the most common plugins used in an ember-cli app provide the cache keys (I believe this is mostly just `ember-cli-htmlbars-inline-precompile` but will test in a few apps).

stefanpenner commented:
> I'd like to get final review on this

Will find time today to review!

rwjblue commented:
Updated again to properly support babel 6's ability to pass options along with the plugin (as an array). Checked off the task item for "basic babel 6 support".

rwjblue commented:
Working on some supporting PR's / issues that can be merged when this lands:

* https://github.com/ember-cli/ember-cli-htmlbars-inline-precompile/issues/34
* https://github.com/rwjblue/babel-plugin-htmlbars-inline-precompile/compare/v0.1.x...rwjblue:cache-keys

stefanpenner commented on README.md:
```diff
@@ -64,3 +64,38 @@ You don't always need this, review which features need the polyfill here: [ES6 F
+
+In order to aid plugin developers in this process, broccoli-babel-transpiler will invoke two methods on a plugin so that it can augment the cache key:
+
+* `cacheKey` - This method can return any string value to be used in the cache key calculation. This is generally only needed when the configuration provided
```
rather then  describing the property with itself

> cacheKey only matters if the plugin affects the cache key

We may want to say:
```
cacheKey is used to describe any runtime information that may want to invalidate the cache. An example of this would be:

* FEATURE flags for defeatureify
* template-compiler.js for htmlbars pre compiler
* options


```

rwjblue replied:
Great point, updated the README.

stefanpenner commented on index.js:
```diff
@@ -111,6 +116,51 @@ Babel.prototype.optionsHash = function() {
+        if (Array.isArray(plugin)) {
+          // when an array, the second argument is an options hash
+          pluginRequiresBaseDir = false;
+          hash.plugins.push(stringify(plugin[1]));
```
so in the array case, each member of the array is a 100% serializable "pure" object?

rwjblue replied:
Yes, that is my understanding.  But to be clear this is for basic babel 6 support, but I am not 100% sure what is allowed and disallowed in the options provided to a plugin. 

I'm happy to follow this up with a recursive algo that does the same treatment to each key in this second arg also, if we think that is needed.

In the meantime, for babel 5 support, we don't even hit this branch.

stefanpenner replied:
@rwjblue I would prefer we don't accidentally leak something in, so we should error, disable caching or

```
var plugins = options.plugins.slice();
// loop over every plugin


if (Array.isArray(plugin)) {
  // append array like things to the existing loop, so we "flatten" as we iterate
  plugin.forEach(plugin => plugins.push(plugin))
}
```

rwjblue replied:
Sounds good.

rwjblue replied:
Updated to recursively add items to the array being iterated (and simplified the logic in general).

stefanpenner commented on index.js:
```diff
@@ -111,6 +116,51 @@ Babel.prototype.optionsHash = function() {
+        } else if (pluginType === 'string') {
+          pluginRequiresBaseDir = false;
+          hash.plugins.push(plugin);
+        }
```
What happens when plugin is not and array, function object or string rather undefined symbol number boolean or some funky host object.

rwjblue replied:
Then `pluginRequiresBaseDir` is true, and the warning below is issued.  I'm not 100% sure what _to do_ about these other types of objects, but I feel that we need to get issues reported or something so we can handle those cases...

stefanpenner replied:
ok, that makes sense. It was just a-bit unclear as written. (symmetry of branches is an easy way to see if all cases are handled, and what happens in each case).

rwjblue replied:
Yeah, I refactored this to ensure there is an `else` that ensures the cache is invalidated (date + rand).  That should make it much simpler to grok I think....

rwjblue commented:
@stefanpenner - Ready for another round. Sorry for so much review work here, turns out cache invalidate is not easy....

rtablada commented:
This is awesomely exciting! Especially for starting developers who may not understand the file not found errors or developers with rotational drives or slower machines where the cost of a full build restart is tough

stefanpenner commented:
> Especially for starting developers who may not understand the file not found errors

seems unrelated, if this is happening please open an issue.

> developers with rotational drives or slower machines where the cost of a full build restart is tough

yes, this will prevent un-needed manual purging of the persistent, which should help keep the work flows performant.

rtablada commented:
@stefanpenner cool, my students have been using 2.6 this time around so final projects they'll be using 2.7 so we'll see if theres still a ton of file not founds. I think there was an issue with cache busting b/c all the time broccoli thought files were there that weren't...

stefanpenner commented:
> I think there was an issue with cache busting b/c all the time broccoli thought files were there that weren't...

This is not a known bug, please open a new issue and report it. It sounds like something we should investigate.
